DataStructure/DSU_rollback.h

```diff
@@ -8,36 +12,37 @@ struct DSU {
     vector<Data> change;
 
     void init(int n) {
-        par = vector<int> (n + 5, 0);
-        REP(i, n+5) par[i] = i;
+        par = vector<int> (n + 5, -1);
         change.clear();
     }
 
     // find root of x.
-    // t should be current time. We use it to update `change`.
-    int getRoot(int x, int t) {
-        if (par[x] != x) {
-            change.push_back({t, x, par[x]});
-            par[x] = getRoot(par[x], t);
-        }
-        return par[x];
+    // if par[x] < 0 then x is a root, and its tree has -par[x] nodes
+    int getRoot(int x) {
```

Comment:
Why no path compression?

Reply:
I would not use path compression with rollback. Imagine when you have a path, then make some changes (compress the path too), then roll back to the state before you compressed the path. It would compress a path many times. So it might not too optimized.